source/iterator/cdc.go

```diff
@@ -0,0 +1,487 @@
+		return false, fmt.Errorf("load rows: %w", err)
+	}
+
+	return false, nil
+}
+
+// Next get new record.
+// nolint:funlen,nolintlint
+func (i *cdcIterator) Next(ctx context.Context) (sdk.Record, error) {
+	row := make(map[string]any)
+	if err := i.rows.MapScan(row); err != nil {
+		return sdk.Record{}, fmt.Errorf("scan rows: %w", err)
+	}
+
+	transformedRow, err := columntypes.TransformRow(ctx, row, i.columnTypes)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("transform row column types: %w", err)
+	}
+
+	id, ok := transformedRow[columnTrackingID].(int64)
+	if !ok {
+		return sdk.Record{}, ErrWrongTrackingIDType
+	}
+
+	operationTypeBt, ok := transformedRow[columnOperationType].([]byte)
+	if !ok {
+		return sdk.Record{}, ErrWrongTrackingOperatorType
+	}
+
+	pos := position.Position{
+		IteratorType: position.TypeCDC,
+		CDCLastID:    int(id),
+		SuffixName:   i.trackingTable[len(i.trackingTable)-6:],
+	}
+
+	convertedPosition, err := pos.ConvertToSDKPosition()
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("convert position %w", err)
+	}
+
+	keysMap := make(map[string]any)
+	for _, val := range i.keys {
+		if _, ok := transformedRow[val]; !ok {
+			return sdk.Record{}, fmt.Errorf("key %v, %w", val, ErrNoKey)
+		}
+		keysMap[val] = transformedRow[val]
+	}
+
+	delete(transformedRow, columnOperationType)
+	delete(transformedRow, columnTrackingID)
+
+	transformedRowBytes, err := json.Marshal(transformedRow)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("marshal row: %w", err)
+	}
+
+	i.position = &pos
+	metadata := sdk.Metadata(map[string]string{metadataTable: i.table})
+	metadata.SetCreatedAt(time.Now())
+
+	switch actionType(operationTypeBt) {
+	case insertOperation:
+		return sdk.Util.Source.NewRecordCreate(convertedPosition, metadata,
+			sdk.StructuredData(keysMap), sdk.RawData(transformedRowBytes)), nil
+	case updateOperation:
+		return sdk.Util.Source.NewRecordUpdate(convertedPosition, metadata,
+			sdk.StructuredData(keysMap), nil, sdk.RawData(transformedRowBytes)), nil
+	case deleteOperation:
+		return sdk.Util.Source.NewRecordDelete(convertedPosition, metadata,
+			sdk.StructuredData(keysMap)), nil
+	default:
+		return sdk.Record{}, ErrUnknownOperatorType
+	}
+}
+
+// Stop shutdown iterator.
+func (i *cdcIterator) Stop() error {
+	// send signal for finish clear tracking table.
+	i.tableSrv.stopCh <- struct{}{}
+
+	if i.rows != nil {
+		err := i.rows.Close()
+		if err != nil {
+			return fmt.Errorf("close rows: %w", err)
+		}
+	}
+
+	select {
+	// wait until clearing tracking table will be finished.
+	case <-i.tableSrv.canCloseCh:
+		if i.db != nil {
+			i.tableSrv.close()
+
+			err := i.db.Close()
+			if err != nil {
+				return fmt.Errorf("close db:%w", err)
+			}
+		}
+	// waiting timeout.
+	case <-time.After(waitingTimeoutSec * time.Second):
+		if i.db != nil {
+			i.tableSrv.close()
+
+			err := i.db.Close()
+			if err != nil {
+				return fmt.Errorf("close db:%w", err)
+			}
+
+			return nil
+		}
+	}
```

Comment:
The only difference I see between these two cases is `return nil` which we have when the timeout is reached.

Maybe we can simplify the switch as:
```go
select {
    	case <-i.tableSrv.canCloseCh:
    	case <-time.After(waitingTimeoutSec * time.Second):
            // close tableSrv
}
```

Reply:
For us is valid case when we get signal from this channel tableSrv.canCloseCh, because it is means that connector clears tracking table and we can safely close it. We use timer for purpose if clearing tracking process stuck

Reply:
Added comments and logs